default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+
+    def print_production_cost(self):
+        """Print production_cost cache."""
+        universe = fo.getUniverse()
+        print "Cached production cost per planet:"
+        for pid in self.production_cost:
+            print "  %s:" % universe.getPlanet(pid).name, self.production_cost[pid]
+
+    def print_production_time(self):
+        """Print production_time cache."""
+        universe = fo.getUniverse()
+        print "Cached production cost per planet:"
+        for pid in self.production_time:
+            print "  %s:" % universe.getPlanet(pid).name, self.production_time[pid]
+
+    def print_all(self):
+        """Print the entire ship design cache."""
+        print
+        print "Printing the ShipDesignAI cache..."
+        self.print_testhulls()
+        self.print_design_id_by_name()
+        self.print_part_by_partname()
+        self.print_strictly_worse_parts()
+        self.print_map_reference_design_name()
+        self.print_hulls_for_planets()
+        self.print_parts_for_planets()
+        self.print_best_designs()
+        self.print_production_cost()
+        self.print_production_time()
+        print "-----"
+        print
+
+    def _update_cost_cache(self):
+        """Cache the production cost and time for each part and hull for each planet (with shipyard) for this turn."""
+        self.production_cost.clear()
+        self.production_time.clear()
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        for pid in _get_planets_with_shipyard():
+            self.production_time[pid] = {}
+            self.production_cost[pid] = {}
+            for partname in empire.availableShipParts:
+                part = _get_part_type(partname)
+                self.production_cost[pid][partname] = part.productionCost(empire_id, pid)
+                self.production_time[pid][partname] = part.productionTime(empire_id, pid)
+            for hullname in empire.availableShipHulls:
+                hull = fo.getHullType(hullname)
+                self.production_cost[pid][hullname] = hull.productionCost(empire_id, pid)
+                self.production_time[pid][hullname] = hull.productionTime(empire_id, pid)
+
+    def _build_cache_after_load(self):
+        """Build cache after loading or starting a game.
+
+        This function is supposed to be called after a reload or at the first turn.
+        It reads out all the existing ship designs and then updates the following cache:
+        - map_reference_design_name
+        - design_id_by_name
+        """
+        if self.map_reference_design_name or self.design_id_by_name:
+            print "WARNING: In ShipDesignAI.py: Cache._build_cache_after_load() called but cache is not empty."
+        for ID in fo.getEmpire().allShipDesigns:
+            design = fo.getShipDesign(ID)
+            reference_name = _build_reference_name(design.hull, [part for part in design.parts])
```

Comment:
`reference_name = _build_reference_name(design.hull, design.parts)` no need to convert to list list_like object.  In case if you need list just `lsit(design.parts)`